streams/concat_streams.ts

```diff
@@ -1,78 +1,77 @@
  * const stream3 = ReadableStream.from([7, 8, 9])
  *
- * for await (const x of new ConcatStreams([stream1, stream2, stream3]))
+ * for await (const x of concatStreams([stream1, stream2, stream3]))
  *   console.log(x)
  * ```
  */
-export class ConcatStreams<T> extends ReadableStream<T> {
-  /** Constructs new instance. */
-  constructor(
-    streams: AsyncIterable<ReadableStream<T>> | Iterable<ReadableStream<T>>,
-  ) {
-    const gen = async function* () {
-      const iter = Symbol.asyncIterator in streams
-        ? streams[Symbol.asyncIterator]()
-        : streams[Symbol.iterator]();
-      x: while (true) {
-        const { done, value } = await iter.next();
+export function concatStreams<T>(
+  streams: AsyncIterable<ReadableStream<T>> | Iterable<ReadableStream<T>>,
+): ReadableStream<T> {
+  const gen = async function* () {
+    const iter = Symbol.asyncIterator in streams
+      ? streams[Symbol.asyncIterator]()
+      : streams[Symbol.iterator]();
+    x: while (true) {
+      const { done, value } = await iter.next();
+      if (done) {
+        break;
+      }
+      const reader = value.getReader();
+      while (true) {
+        const { done, value } = await reader.read();
         if (done) {
           break;
         }
-        const reader = value.getReader();
-        while (true) {
-          const { done, value } = await reader.read();
-          if (done) {
-            break;
-          }
-          try {
-            yield value;
-          } catch (reason) {
-            const promises = [reader.cancel(reason)];
-            while (true) {
-              const { done, value } = await iter.next();
-              if (done) {
-                break;
-              }
-              promises.push(value.cancel(reason));
+        try {
+          yield value;
+        } catch (reason) {
+          const promises = [reader.cancel(reason)];
+          while (true) {
+            const { done, value } = await iter.next();
+            if (done) {
+              break;
             }
-            await Promise.allSettled(promises);
-            break x;
+            promises.push(value.cancel(reason));
           }
+          await Promise.allSettled(promises);
+          break x;
         }
       }
-    }();
-    let lock = false;
-    super({
+    }
+  }();
+  return new ReadableStream<T>(
+    {
+      lock: false,
       async pull(controller) {
-        while (lock) {
+        while (this.lock) {
           await new Promise((a) => setTimeout(a, 0));
         }
-        lock = true;
+        this.lock = true;
         const { done, value } = await gen.next();
         if (done) {
           controller.close();
         } else {
           controller.enqueue(value);
         }
-        lock = false;
+        this.lock = false;
       },
       async cancel(reason) {
         await gen.throw(reason).catch(() => {});
       },
-    });
-  }
+    } as UnderlyingSource<T> & { lock: boolean },
```

Comment:
Can we please move `lock` into the function body rather than as an extended argument for the `ReadableStream` constructor? I understand your personal preference for declaring variables in the lowest possible scope, but this seems too unnatural for a Standard Library API.

Reply:
Np, it was outside before. I should have just left it out.